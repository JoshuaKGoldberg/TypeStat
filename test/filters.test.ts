import path from "node:path";
import { describe, expect, it } from "vitest";

import { runMutationTest } from "../src/tests/testSetup.js";

describe("filters", () => {
	it("empty", async () => {
		const caseDir = path.join(import.meta.dirname, "./cases/filters/empty");
		const { actualContent, expectedFilePath, options } =
			await runMutationTest(caseDir);
		await expect(actualContent).toMatchFileSnapshot(expectedFilePath);
		expect(options).toMatchSnapshot("options");
	}, 7000);

	it("one", async () => {
		const caseDir = path.join(import.meta.dirname, "./cases/filters/one");
		const { actualContent, expectedFilePath, options } =
			await runMutationTest(caseDir);
		await expect(actualContent).toMatchFileSnapshot(expectedFilePath);
		expect(options).toMatchSnapshot("options");
	});

	it("two", async () => {
		const caseDir = path.join(import.meta.dirname, "./cases/filters/two");
		const { actualContent, expectedFilePath, options } =
			await runMutationTest(caseDir);
		await expect(actualContent).toMatchFileSnapshot(expectedFilePath);
		expect(options).toMatchSnapshot("options");
	});
});
