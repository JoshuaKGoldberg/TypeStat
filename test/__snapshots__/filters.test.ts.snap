// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`filters > empty > options 1`] = `
"[
  {
    "cleanups": {
      "suppressTypeErrors": true
    },
    "files": {
      "above": "",
      "below": "",
      "renameExtensions": false
    },
    "filters": [],
    "fixes": {
      "importExtensions": false,
      "incompleteTypes": true,
      "missingProperties": false,
      "noImplicitAny": true,
      "noImplicitThis": false,
      "noInferableTypes": false,
      "strictNonNullAssertions": false
    },
    "hints": {
      "react": {
        "propTypes": "whenRequired",
        "propTypesOptionality": "asWritten"
      }
    },
    "include": [
      "<rootDir>/actual.ts"
    ],
    "mutators": [
      [
        "fixImportExtensions",
        null
      ],
      [
        "fixIncompleteTypes",
        null
      ],
      [
        "fixMissingProperties",
        null
      ],
      [
        "fixNoImplicitAny",
        null
      ],
      [
        "fixNoImplicitThis",
        null
      ],
      [
        "fixNoInferableTypes",
        null
      ],
      [
        "fixStrictNonNullAssertions",
        null
      ]
    ],
    "output": {},
    "package": {
      "directory": "<rootDir>",
      "file": "<rootDir>/package.json"
    },
    "parsedTsConfig": {
      "options": {
        "configFilePath": "<rootDir>/tsconfig.json",
        "noEmit": true,
        "noImplicitAny": true,
        "noImplicitThis": false,
        "strictNullChecks": true
      },
      "fileNames": [
        "<rootDir>/actual.ts"
      ],
      "typeAcquisition": {
        "enable": false,
        "include": [],
        "exclude": []
      },
      "raw": {
        "files": [
          "actual.ts"
        ]
      },
      "errors": [],
      "wildcardDirectories": {},
      "compileOnSave": false
    },
    "postProcess": {
      "shell": []
    },
    "projectPath": "<rootDir>/tsconfig.json",
    "types": {
      "strictNullChecks": true
    }
  }
]"
`;

exports[`filters > empty > output 1`] = `
"[log] Preparing language services to visit files...
[log] Prepared language services for 1 files...
[stdout] Starting the core mutation engine. This terminal will log whenever a "wave" of mutations are written to files.
[stdout] Core mutations will complete when two waves pass with no mutations.
[stdout] The following fixes will be applied:
[stdout] 	* incompleteTypes
[stdout] 	* noImplicitAny

[stdout] Starting wave 1...
[log] fixIncompleteReturnTypes found 2 mutations (2 insertions) in <rootDir>/actual.ts: [mutations]
[log] fixIncompleteTypes found 2 mutations (2 insertions) in <rootDir>/actual.ts: [mutations]
[stdout] Completed wave 1. Wrote mutations to 1 file.
[log] Preparing language services to visit files...
[log] Prepared language services for 1 files...
[stdout] Starting wave 2...
[stdout] Completed wave 2. Wrote mutations to 0 files.
[stdout] Done.

[stdout] Applying post-fix cleanups...
[stdout] Done.
"
`;

exports[`filters > one > options 1`] = `
"[
  {
    "cleanups": {
      "suppressTypeErrors": true
    },
    "files": {
      "above": "",
      "below": "",
      "renameExtensions": false
    },
    "filters": [
      "FunctionDeclaration[name.text=one]"
    ],
    "fixes": {
      "importExtensions": false,
      "incompleteTypes": true,
      "missingProperties": false,
      "noImplicitAny": true,
      "noImplicitThis": false,
      "noInferableTypes": false,
      "strictNonNullAssertions": false
    },
    "hints": {
      "react": {
        "propTypes": "whenRequired",
        "propTypesOptionality": "asWritten"
      }
    },
    "include": [
      "<rootDir>/actual.ts"
    ],
    "mutators": [
      [
        "fixImportExtensions",
        null
      ],
      [
        "fixIncompleteTypes",
        null
      ],
      [
        "fixMissingProperties",
        null
      ],
      [
        "fixNoImplicitAny",
        null
      ],
      [
        "fixNoImplicitThis",
        null
      ],
      [
        "fixNoInferableTypes",
        null
      ],
      [
        "fixStrictNonNullAssertions",
        null
      ]
    ],
    "output": {},
    "package": {
      "directory": "<rootDir>",
      "file": "<rootDir>/package.json"
    },
    "parsedTsConfig": {
      "options": {
        "configFilePath": "<rootDir>/tsconfig.json",
        "noEmit": true,
        "noImplicitAny": true,
        "noImplicitThis": false,
        "strictNullChecks": true
      },
      "fileNames": [
        "<rootDir>/actual.ts"
      ],
      "typeAcquisition": {
        "enable": false,
        "include": [],
        "exclude": []
      },
      "raw": {
        "files": [
          "actual.ts"
        ]
      },
      "errors": [],
      "wildcardDirectories": {},
      "compileOnSave": false
    },
    "postProcess": {
      "shell": []
    },
    "projectPath": "<rootDir>/tsconfig.json",
    "types": {
      "strictNullChecks": true
    }
  }
]"
`;

exports[`filters > one > output 1`] = `
"[log] Preparing language services to visit files...
[log] Prepared language services for 1 files...
[stdout] Starting the core mutation engine. This terminal will log whenever a "wave" of mutations are written to files.
[stdout] Core mutations will complete when two waves pass with no mutations.
[stdout] The following fixes will be applied:
[stdout] 	* incompleteTypes
[stdout] 	* noImplicitAny

[stdout] Starting wave 1...
[log] fixIncompleteReturnTypes found 1 mutations (1 insertions) in <rootDir>/actual.ts: [mutations]
[log] fixIncompleteTypes found 1 mutations (1 insertions) in <rootDir>/actual.ts: [mutations]
[stdout] Completed wave 1. Wrote mutations to 1 file.
[log] Preparing language services to visit files...
[log] Prepared language services for 1 files...
[stdout] Starting wave 2...
[stdout] Completed wave 2. Wrote mutations to 0 files.
[stdout] Done.

[stdout] Applying post-fix cleanups...
[log] suppressTypeIssues found 1 mutations (1 insertions) in <rootDir>/actual.ts: [mutations]
[stdout] Done.
"
`;

exports[`filters > two > options 1`] = `
"[
  {
    "cleanups": {
      "suppressTypeErrors": true
    },
    "files": {
      "above": "",
      "below": "",
      "renameExtensions": false
    },
    "filters": [
      "MethodDeclaration[name.text=dispose]",
      "CallExpression[expression.text=teardown]"
    ],
    "fixes": {
      "importExtensions": false,
      "incompleteTypes": true,
      "missingProperties": false,
      "noImplicitAny": true,
      "noImplicitThis": false,
      "noInferableTypes": false,
      "strictNonNullAssertions": false
    },
    "hints": {
      "react": {
        "propTypes": "whenRequired",
        "propTypesOptionality": "asWritten"
      }
    },
    "include": [
      "<rootDir>/actual.ts"
    ],
    "mutators": [
      [
        "fixImportExtensions",
        null
      ],
      [
        "fixIncompleteTypes",
        null
      ],
      [
        "fixMissingProperties",
        null
      ],
      [
        "fixNoImplicitAny",
        null
      ],
      [
        "fixNoImplicitThis",
        null
      ],
      [
        "fixNoInferableTypes",
        null
      ],
      [
        "fixStrictNonNullAssertions",
        null
      ]
    ],
    "output": {},
    "package": {
      "directory": "<rootDir>",
      "file": "<rootDir>/package.json"
    },
    "parsedTsConfig": {
      "options": {
        "configFilePath": "<rootDir>/tsconfig.json",
        "noEmit": true,
        "noImplicitAny": true,
        "noImplicitThis": false,
        "strictNullChecks": true
      },
      "fileNames": [
        "<rootDir>/actual.ts"
      ],
      "typeAcquisition": {
        "enable": false,
        "include": [],
        "exclude": []
      },
      "raw": {
        "files": [
          "actual.ts"
        ]
      },
      "errors": [],
      "wildcardDirectories": {},
      "compileOnSave": false
    },
    "postProcess": {
      "shell": []
    },
    "projectPath": "<rootDir>/tsconfig.json",
    "types": {
      "strictNullChecks": true
    }
  }
]"
`;

exports[`filters > two > output 1`] = `
"[log] Preparing language services to visit files...
[log] Prepared language services for 1 files...
[stdout] Starting the core mutation engine. This terminal will log whenever a "wave" of mutations are written to files.
[stdout] Core mutations will complete when two waves pass with no mutations.
[stdout] The following fixes will be applied:
[stdout] 	* incompleteTypes
[stdout] 	* noImplicitAny

[stdout] Starting wave 1...
[stdout] Completed wave 1. Wrote mutations to 0 files.
[stdout] Done.

[stdout] Applying post-fix cleanups...
[log] suppressTypeIssues found 2 mutations (2 insertions) in <rootDir>/actual.ts: [mutations]
[stdout] Done.
"
`;
